lib/FreeRTOS-Plus-TCP/source/FreeRTOS_TCP_IP.c

```diff
@@ -1158,161 +1172,179 @@ UBaseType_t uxNewMSS;
-			endian number. */
-			uxNewMSS = usChar2u16( pucPtr + 2 );
+		#if( ipconfigUSE_TCP_WIN != 1 )
+			/* Without scaled windows, MSS is the only interesting option. */
+			return LOOP_BREAK;
+		#else
+			/* Or else we continue to check another option: selective ACK. */
+			( *pucPtr ) += TCP_OPT_MSS_LEN;
+		#endif	/* ipconfigUSE_TCP_WIN != 1 */
+	}
+	else
+	{
+		/* All other options have a length field, so that we easily
+		can skip past them. */
+		unsigned char len = ( *pucPtr )[ 1 ];
+		if( ( len < 2 ) || ( len > xRemainingOptionsBytes ) )
+		{
+			/* If the length field is too small or too big, the options are malformed.
+			Don't process them further. */
+			return LOOP_BREAK;
+		}
 
-			if( pxSocket->u.xTCP.usInitMSS != uxNewMSS )
+		#if( ipconfigUSE_TCP_WIN == 1 )
+		{
+			/* Selective ACK: the peer has received a packet but it is missing earlier
+			packets.  At least this packet does not need retransmission anymore
+			ulTCPWindowTxSack( ) takes care of this administration. */
+			if( ( *pucPtr )[0] == TCP_OPT_SACK_A )
 			{
-				/* Perform a basic check on the the new MSS. */
-				if( uxNewMSS == 0 )
-				{
-					break;
-				}
-
-				FreeRTOS_debug_printf( ( "MSS change %u -> %lu\n", pxSocket->u.xTCP.usInitMSS, uxNewMSS ) );
-			}
+				len -= 2;
+				( *pucPtr ) += 2;
 
-			if( pxSocket->u.xTCP.usInitMSS > uxNewMSS )
-			{
-				/* our MSS was bigger than the MSS of the other party: adapt it. */
-				pxSocket->u.xTCP.bits.bMssChange = pdTRUE_UNSIGNED;
-				if( ( pxTCPWindow != NULL ) && ( pxSocket->u.xTCP.usCurMSS > uxNewMSS ) )
+				while( len >= 8 )
 				{
-					/* The peer advertises a smaller MSS than this socket was
-					using.  Use that as well. */
-					FreeRTOS_debug_printf( ( "Change mss %d => %lu\n", pxSocket->u.xTCP.usCurMSS, uxNewMSS ) );
-					pxSocket->u.xTCP.usCurMSS = ( uint16_t ) uxNewMSS;
+					prvCheckOptions_innerLoop(pucPtr, pxSocket, &len);
 				}
-				pxTCPWindow->xSize.ulRxWindowLength = ( ( uint32_t ) uxNewMSS ) * ( pxTCPWindow->xSize.ulRxWindowLength / ( ( uint32_t ) uxNewMSS ) );
-				pxTCPWindow->usMSSInit = ( uint16_t ) uxNewMSS;
-				pxTCPWindow->usMSS = ( uint16_t ) uxNewMSS;
-				pxSocket->u.xTCP.usInitMSS = ( uint16_t ) uxNewMSS;
-				pxSocket->u.xTCP.usCurMSS = ( uint16_t ) uxNewMSS;
+				/* len should be 0 by now. */
 			}
-
-			#if( ipconfigUSE_TCP_WIN != 1 )
-				/* Without scaled windows, MSS is the only interesting option. */
-				break;
-			#else
-				/* Or else we continue to check another option: selective ACK. */
-				pucPtr += TCP_OPT_MSS_LEN;
-			#endif	/* ipconfigUSE_TCP_WIN != 1 */
 		}
-		else
+		#endif	/* ipconfigUSE_TCP_WIN == 1 */
+
+		( *pucPtr ) += len;
+	}
+	return LOOP_CONTINUE;
+}
+
+
+/*-----------------------------------------------------------*/
+
+static FORCE_INLINE void prvCheckOptions_innerLoop( const unsigned char ** const pucPtr, FreeRTOS_Socket_t ** const pxSocket, unsigned char * const pucLen)
+{
+uint32_t ulFirst = ulChar2u32( ( *pucPtr ) );
```

Comment:
Assumes that pucPtr is long enough to hold a valid SACK. Maybe at least say that in a comment? Better, since you're passing the length anyway, add an explicit check. Yes, the latter is arguably redundant, but this is the kind of code that calls for paranoia.

Reply:
Agreed, but could I leave this for a separate PR? Two reasons:
- This PR contains only non-functional changes, I think it's cleaner to keep it that way
- If I add new assertions, CBMC will check them :) I'd like to get the proofs running on the code we already have before I start sprinkling assertions everywhere, it would be useful to start from a known-good state (where Mark left off)